File: frontend/src/components/benchmark/content/clustering_coefficient/implementations/clustering_coefficient.js
class ClusteringCoefficient {
    constructor() {
        this.adjacency = new Map();
        this.nodes = new Set();
    }

    addEdge(u, v) {
        if (!this.adjacency.has(u)) {
            this.adjacency.set(u, new Set());
        }
        if (!this.adjacency.has(v)) {
            this.adjacency.set(v, new Set());
        }
        this.adjacency.get(u).add(v);
        this.adjacency.get(v).add(u);
        this.nodes.add(u);
        this.nodes.add(v);
    }

    buildFromEdges(edges) {
        for (const [u, v] of edges) {
            this.addEdge(u, v);
        }
    }

    localClusteringCoefficient(node) {
        if (!this.adjacency.has(node)) {
            return 0.0;
        }

        const neighbors = this.adjacency.get(node);
        const degree = neighbors.size;

        if (degree < 2) {
            return 0.0;
        }

        let triangles = 0;
        const neighborList = Array.from(neighbors);

        for (let i = 0; i < neighborList.length; i++) {
            for (let j = i + 1; j < neighborList.length; j++) {
                const u = neighborList[i];
                const v = neighborList[j];
                if (this.adjacency.get(u).has(v)) {
                    triangles++;
                }
            }
        }

        const maxPossible = degree * (degree - 1) / 2;
        return triangles / maxPossible;
    }

    allLocalClusteringCoefficients() {
        const result = {};
        for (const node of this.nodes) {
            result[node] = this.localClusteringCoefficient(node);
        }
        return result;
    }

    averageClusteringCoefficient() {
        if (this.nodes.size === 0) {
            return 0.0;
        }

        let total = 0.0;
        for (const node of this.nodes) {
            total += this.localClusteringCoefficient(node);
        }
        return total / this.nodes.size;
    }

    globalClusteringCoefficient() {
        let triangles = 0;
        let triplets = 0;

        for (const node of this.nodes) {
            const degree = this.adjacency.get(node).size;
            if (degree >= 2) {
                triplets += degree * (degree - 1) / 2;
            }

            const neighborList = Array.from(this.adjacency.get(node));
            for (let i = 0; i < neighborList.length; i++) {
                for (let j = i + 1; j < neighborList.length; j++) {
                    const u = neighborList[i];
                    const v = neighborList[j];
                    if (this.adjacency.get(u).has(v)) {
                        triangles++;
                    }
                }
            }
        }

        if (triplets === 0) {
            return 0.0;
        }

        return triangles / triplets;
    }
}

module.exports = ClusteringCoefficient;

File: frontend/src/components/benchmark/content/clustering_coefficient/implementations/clustering_coefficient.go
package main

type ClusteringCoefficient struct {
    adjacency map[int]map[int]bool
    nodes     map[int]bool
}

func NewClusteringCoefficient() *ClusteringCoefficient {
    return &ClusteringCoefficient{
        adjacency: make(map[int]map[int]bool),
        nodes:     make(map[int]bool),
    }
}

func (cc *ClusteringCoefficient) AddEdge(u, v int) {
    if cc.adjacency[u] == nil {
        cc.adjacency[u] = make(map[int]bool)
    }
    if cc.adjacency[v] == nil {
        cc.adjacency[v] = make(map[int]bool)
    }
    cc.adjacency[u][v] = true
    cc.adjacency[v][u] = true
    cc.nodes[u] = true
    cc.nodes[v] = true
}

func (cc *ClusteringCoefficient) BuildFromEdges(edges [][2]int) {
    for _, edge := range edges {
        cc.AddEdge(edge[0], edge[1])
    }
}

func (cc *ClusteringCoefficient) LocalClusteringCoefficient(node int) float64 {
    if cc.adjacency[node] == nil {
        return 0.0
    }

    neighbors := cc.adjacency[node]
    degree := len(neighbors)

    if degree < 2 {
        return 0.0
    }

    triangles := 0
    neighborList := []int{}
    for neighbor := range neighbors {
        neighborList = append(neighborList, neighbor)
    }

    for i := 0; i < len(neighborList); i++ {
        for j := i + 1; j < len(neighborList); j++ {
            u := neighborList[i]
            v := neighborList[j]
            if cc.adjacency[u][v] {
                triangles++
            }
        }
    }

    maxPossible := degree * (degree - 1) / 2
    return float64(triangles) / float64(maxPossible)
}

func (cc *ClusteringCoefficient) AllLocalClusteringCoefficients() map[int]float64 {
    result := make(map[int]float64)
    for node := range cc.nodes {
        result[node] = cc.LocalClusteringCoefficient(node)
    }
    return result
}

func (cc *ClusteringCoefficient) AverageClusteringCoefficient() float64 {
    if len(cc.nodes) == 0 {
        return 0.0
    }

    total := 0.0
    for node := range cc.nodes {
        total += cc.LocalClusteringCoefficient(node)
    }
    return total / float64(len(cc.nodes))
}

func (cc *ClusteringCoefficient) GlobalClusteringCoefficient() float64 {
    triangles := 0
    triplets := 0

    for node := range cc.nodes {
        degree := len(cc.adjacency[node])
        if degree >= 2 {
            triplets += degree * (degree - 1) / 2
        }

        neighborList := []int{}
        for neighbor := range cc.adjacency[node] {
            neighborList = append(neighborList, neighbor)
        }

        for i := 0; i < len(neighborList); i++ {
            for j := i + 1; j < len(neighborList); j++ {
                u := neighborList[i]
                v := neighborList[j]
                if cc.adjacency[u][v] {
                    triangles++
                }
            }
        }
    }

    if triplets == 0 {
        return 0.0
    }

    return float64(triangles) / float64(triplets)
}